File: sdk-android/src/main/java/cu/suitetecsa/sdk/android/ContactsCollectorImpl.java
package cu.suitetecsa.sdk.android;

import android.annotation.SuppressLint;
import android.content.Context;
import android.database.Cursor;
import android.provider.ContactsContract;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;

import cu.suitetecsa.sdk.android.model.Contact;

/**
 * Implementation of {@link ContactsCollector} for collecting contact information from the device.
 */
class ContactsCollectorImpl implements ContactsCollector {
    private final Context context;

    /**
     * Constructor for {@link ContactsCollectorImpl}.
     *
     * @param context the {@link Context} used to access the contacts.
     */
    ContactsCollectorImpl(Context context) {
        this.context = context;
    }

    /**
     * Collects contact information from the device.
     * <p>
     * Note: This method requires {@link android.Manifest.permission#READ_CONTACTS} permission.
     * Ensure the permission is granted before invoking this method.
     * </p>
     *
     * @return A list of {@link Contact} objects containing contact names and their associated phone numbers.
     */
    @Override
    public List<Contact> collect() {
        Map<String, Contact> contactsMap = new HashMap<>();

        Cursor cursor = context.getContentResolver().query(
                ContactsContract.CommonDataKinds.Phone.CONTENT_URI,
                new String[]{
                        ContactsContract.CommonDataKinds.Phone.DISPLAY_NAME,
                        ContactsContract.CommonDataKinds.Phone.NUMBER,
                        ContactsContract.Contacts.PHOTO_URI
                },
                null,
                null,
                ContactsContract.CommonDataKinds.Phone.DISPLAY_NAME + " ASC");

        if (cursor != null && cursor.moveToFirst()) {
            while (!cursor.isAfterLast()) {
                @SuppressLint("Range") String name = cursor.getString(cursor.getColumnIndex(ContactsContract.CommonDataKinds.Phone.DISPLAY_NAME));
                @SuppressLint("Range") String phoneNumber = cursor.getString(cursor.getColumnIndex(ContactsContract.CommonDataKinds.Phone.NUMBER));
                @SuppressLint("Range") String photoUri = cursor.getString(cursor.getColumnIndex(ContactsContract.CommonDataKinds.Phone.PHOTO_URI));

                if (!contactsMap.containsKey(name)) {
                    contactsMap.put(name, new Contact(name, phoneNumber, photoUri));
                }

                cursor.moveToNext();
            }
            cursor.close();
        }
        return new ArrayList<>(contactsMap.values());
    }
}

File: sdk-android/src/main/java/io/github/suitetecsa/sdk/android/ContactsCollectorImpl.kt
package io.github.suitetecsa.sdk.android

import android.content.ContentResolver
import android.content.Context
import android.database.Cursor
import android.provider.ContactsContract
import android.util.Log
import io.github.suitetecsa.sdk.android.model.Contact

/**
 * Implementation of [ContactsCollector] for collecting contact information from the device.
 */
internal class ContactsCollectorImpl(private val context: Context) : ContactsCollector {

    /**
     * Collects contact information from the device.
     *
     * This method retrieves contact names, phone numbers, and photo URIs from the device's
     * contacts provider.
     *
     * Note: This method requires the [android.Manifest.permission.READ_CONTACTS] permission.
     * Ensure the permission is granted before invoking this method.
     *
     * @return A list of [Contact] objects, each containing a contact's name, phone number, and
     * optional photo URI. Returns an empty list if no contacts are found or if there's an error.
     */
    @Suppress("TooGenericExceptionCaught", "NestedBlockDepth")
    override fun collect(): List<Contact> {
        val contacts = mutableListOf<Contact>()
        val contentResolver: ContentResolver = context.contentResolver
        var cursor: Cursor? = null

        try {
            cursor = contentResolver.query(
                ContactsContract.CommonDataKinds.Phone.CONTENT_URI,
                PROJECTION,
                null,
                null,
                SORT_ORDER
            )

            cursor?.use { // Use 'use' to ensure cursor is closed automatically
                val nameColumnIndex = it.getColumnIndex(ContactsContract.CommonDataKinds.Phone.DISPLAY_NAME)
                val numberColumnIndex = it.getColumnIndex(ContactsContract.CommonDataKinds.Phone.NUMBER)
                val photoUriColumnIndex = it.getColumnIndex(ContactsContract.CommonDataKinds.Phone.PHOTO_URI)

                if (it.moveToFirst()) {
                    do {
                        val name = it.getStringOrNull(nameColumnIndex) ?: "" // Handle nulls directly
                        val phoneNumber = it.getStringOrNull(numberColumnIndex) ?: "" // Handle nulls directly
                        val photoUri = it.getStringOrNull(photoUriColumnIndex) // Already nullable

                        if (name.isNotBlank()) {
                            contacts.add(Contact(name, phoneNumber, photoUri))
                        }
                    } while (it.moveToNext())
                }
            }
        } catch (e: SecurityException) {
            // Handle specific permission exception
            Log.e(TAG, "Missing READ_CONTACTS permission", e)
        } catch (e: Exception) {
            Log.e(TAG, "Error collecting contacts", e)
        }
        return contacts.distinctBy { it.name } // Ensure uniqueness based on contact name
    }

    companion object {
        private const val TAG = "ContactsCollector" // Add tag
        private val PROJECTION: Array<String> = arrayOf(
            ContactsContract.CommonDataKinds.Phone.DISPLAY_NAME,
            ContactsContract.CommonDataKinds.Phone.NUMBER,
            ContactsContract.CommonDataKinds.Phone.PHOTO_URI
        )
        private const val SORT_ORDER = "${ContactsContract.CommonDataKinds.Phone.DISPLAY_NAME} ASC"
    }
}

// Extension function to simplify cursor string retrieval
fun Cursor.getStringOrNull(columnIndex: Int): String? {
    return if (columnIndex != -1) getString(columnIndex) else null
}
